File: 3470.permutations-iv.py
#
# @lc app=leetcode id=3470 lang=python3
#
# [3470] Permutations IV
#

# @lc code=start
from typing import List

class Solution:
    def permute(self, n: int, k: int) -> List[int]:
        if n == 0:
            return []
        avail_odd = list(range(1, n + 1, 2))
        avail_even = list(range(2, n + 1, 2))
        fact = [1] * 101
        for i in range(1, 101):
            fact[i] = fact[i - 1] * i
        result = []
        current_k = k
        # First position: try numbers 1 to n in lex order
        picked = False
        for num in range(1, n + 1):
            is_odd_num = (num % 2 == 1)
            ro = len(avail_odd) - (1 if is_odd_num else 0)
            re = len(avail_even) - (1 if not is_odd_num else 0)
            rem_len = n - 1
            next_req_odd = not is_odd_num
            req_odd_count = (rem_len + 1) // 2 if next_req_odd else rem_len // 2
            count = fact[ro] * fact[re] if ro >= 0 and re >= 0 and ro == req_odd_count else 0
            if count >= current_k:
                result.append(num)
                if is_odd_num:
                    avail_odd.remove(num)
                else:
                    avail_even.remove(num)
                last_req_odd = next_req_odd
                picked = True
                break
            current_k -= count
        if not picked:
            return []
        # Remaining positions
        while len(result) < n:
            if last_req_odd:
                cand_list = avail_odd
                is_cand_odd = True
            else:
                cand_list = avail_even
                is_cand_odd = False
            if not cand_list:
                return []
            picked = False
            for i in range(len(cand_list)):
                num = cand_list[i]
                ro = len(avail_odd) - (1 if is_cand_odd else 0)
                re = len(avail_even) - (1 if not is_cand_odd else 0)
                rem_len = len(avail_odd) + len(avail_even) - 1
                next_req_odd = not is_cand_odd
                req_odd_count = (rem_len + 1) // 2 if next_req_odd else rem_len // 2
                count = fact[ro] * fact[re] if ro >= 0 and re >= 0 and ro == req_odd_count else 0
                if count >= current_k:
                    result.append(num)
                    del cand_list[i]
                    last_req_odd = next_req_odd
                    picked = True
                    break
                current_k -= count
            if not picked:
                return []
        return result

# @lc code=end

File: 3470.permutations-iv.cpp
#
# @lc app=leetcode id=3470 lang=cpp
#
# [3470] Permutations IV
#
# @lc code=start
class Solution {
public:
    vector<int> permute(int n, long long k) {
        const long long INF = 2e15;
        
        // Precompute factorials with overflow protection
        vector<long long> fact(n + 1);
        fact[0] = 1;
        for (int i = 1; i <= n; i++) {
            if (fact[i - 1] > INF / i) {
                fact[i] = INF;
            } else {
                fact[i] = fact[i - 1] * i;
            }
        }
        
        // Separate odd and even numbers into sets
        set<int> odd_set, even_set;
        for (int i = 1; i <= n; i++) {
            if (i % 2 == 1) odd_set.insert(i);
            else even_set.insert(i);
        }
        
        vector<int> result;
        int start_parity = -1; // -1: not determined, 0: even, 1: odd
        
        for (int pos = 0; pos < n; pos++) {
            vector<int> candidates;
            
            if (start_parity == -1) {
                // First position: try all numbers in sorted order
                for (int c : odd_set) candidates.push_back(c);
                for (int c : even_set) candidates.push_back(c);
                sort(candidates.begin(), candidates.end());
            } else {
                // Later positions: only try numbers with required parity
                int req_parity = (start_parity + pos) % 2;
                if (req_parity == 1) {
                    for (int c : odd_set) candidates.push_back(c);
                } else {
                    for (int c : even_set) candidates.push_back(c);
                }
            }
            
            bool found = false;
            for (int c : candidates) {
                int c_parity = c % 2;
                int sp = (pos == 0) ? c_parity : start_parity;
                
                // Calculate remaining counts after placing c
                int new_odd = (int)odd_set.size() - (c_parity == 1 ? 1 : 0);
                int new_even = (int)even_set.size() - (c_parity == 0 ? 1 : 0);
                
                // Count positions needing odd/even from pos+1 to n-1
                int odd_needed = 0, even_needed = 0;
                for (int i = pos + 1; i < n; i++) {
                    if ((sp + i) % 2 == 1) odd_needed++;
                    else even_needed++;
                }
                
                // Calculate number of valid permutations
                long long count = 0;
                if (odd_needed == new_odd && even_needed == new_even) {
                    long long a = fact[new_odd], b = fact[new_even];
                    if (a <= INF / b) {
                        count = a * b;
                    } else {
                        count = INF;
                    }
                }
                
                if (count >= k) {
                    result.push_back(c);
                    if (c_parity == 1) odd_set.erase(c);
                    else even_set.erase(c);
                    if (pos == 0) start_parity = c_parity;
                    found = true;
                    break;
                } else {
                    k -= count;
                }
            }
            
            if (!found) return {};
        }
        
        return result;
    }
};
# @lc code=end